### PR TITLE
Factor out ParticleList

### PR DESCRIPTION
**Pull request by fweik**
Follow up on #3251.

Description of changes:
 - Pulling `ParticleList` out of `particle_data.hpp` to get better
   header disentanglement.

### CODE REVIEW

File: src/core/ParticleList.hpp [hunk 1]
```diff
@@ -0,0 +1,58 @@
+#ifndef ESPRESSO_CORE_PARTICLE_LIST_HPP
```
Comment (jngrad) [inline src/core/ParticleList.hpp:21]: missing license header

**Comment by jngrad** [comment]
Could you please rebase the branch, possibly starting from 34c6bd7? It's hard to follow the changes.

**Comment by fweik** [comment]
I'll tell you when it's ready, I first have to remove the regression that I introduced.

**Comment by jngrad** [comment]
By rebasing, I meant squashing all the "fix..." commits together until each commit is self-contained :)

**Comment by fweik** [comment]
@jngrad is this better now?

**Review by jngrad — APPROVED** [review]
Thanks!

**Comment by jngrad** [comment]
bors r-
(failed merge on staging)
